clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureBaseClient.groovy

```diff
@@ -54,28 +54,52 @@ public abstract class AzureBaseClient {
     new ApplicationTokenCredentials(clientId, tenantId, secret, AzureEnvironment.AZURE)
   }
 
-  static Object getAzureOps(Closure getOps, String msgRetry, String msgFail, long count = AZURE_ATOMICOPERATION_RETRY) {
-    // The API call might return a timeout exception or some other Azure CloudException that is not the direct result of the operation
-    //   we are trying to execute; retry and if the final retry fails then throw
-    Object result = null
-    long operationRetry = 0
-    while (operationRetry < count) {
-      try {
-        operationRetry ++
-        result = getOps()
-        operationRetry = count
+  /**
+   * Wrap the call to an Azure operation to handle retry attempts
+   * @param operation - Operation to be execute
+   * @param count - number of retry attempts
+   * @return ServiceRespone returned from operation. If response results in 404 then return null
+   */
+  static <T> ServiceResponse<T> executeOp(Closure<ServiceResponse<T>> operation, long count = AZURE_ATOMICOPERATION_RETRY) {
+
+    long retryCount = count > 0 ? 0 : -1
+    long interval = 200
+
+    // Define which exceptions are ok for retrying
+    def canRetry = { Exception e ->
+      Boolean retry = false
+      if (retryCount < count) {
+        if (e.class == CloudException.class) {
+          def code = (e as CloudException).response.code()
+          retry = (code == HttpURLConnection.HTTP_CLIENT_TIMEOUT
+            || (code >= HttpURLConnection.HTTP_INTERNAL_ERROR && code <= HttpURLConnection.HTTP_GATEWAY_TIMEOUT )
+            || code == 429 )
+        } else if (e.class == SocketTimeoutException.class) {
+          //If we get a socket time out try again
+          retry = true
+        }
       }
-      catch (Exception e) {
-        sleep(200)
-        log.warn("${msgRetry}: ${e.message}")
-        if (operationRetry >= count) {
-          log.error(msgFail)
+      retry
+    }
+
+    while (retryCount < count) {
+      try {
+        retryCount++
+        return operation()
+      } catch (Exception e) {
+        if (canRetry(e)) {
+          log.warn("Retrying Azure operation: ${e.message}")
+          sleep(interval * retryCount)
+        } else if (resourceNotFound(e)) {
+          log.warn("Azure resource(s) not found: ${e.message} ")
+          return null
+        }
+        else {
           throw e
         }
       }
     }
-
-    result
+    null
```

Comment:
Agreed but without it, we get a " Not all execution paths return a value" warning. If we are ok with the warning I will remove it.